File: 151051/Maksimova/task 2-3-4-5/publicator/src/main/java/com/example/repository/IssueRepository.java
package com.example.repository;

import com.example.entities.Issue;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.JpaSpecificationExecutor;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;

import java.util.List;

public interface IssueRepository extends JpaRepository<Issue, Long>, JpaSpecificationExecutor<Issue> {
    boolean existsByTitle(String title);

    @Query("SELECT DISTINCT i FROM Issue i " +
            "LEFT JOIN i.stickers s " +
            "WHERE (:stickerNames IS NULL OR s.name IN :stickerNames) " +
            "AND (:stickerIds IS NULL OR s.id IN :stickerIds) " +
            "AND (:editorLogin IS NULL OR i.editor.login = :editorLogin) " +
            "AND (:title IS NULL OR i.title LIKE %:title%) " +
            "AND (:content IS NULL OR i.content LIKE %:content%)")
    List<Issue> findIssuesByParams(@Param("stickerNames") List<String> stickerNames,
                                   @Param("stickerIds") List<Long> stickerIds,
                                   @Param("editorLogin") String editorLogin,
                                   @Param("title") String title,
                                   @Param("content") String content);

    Page<Issue> findAll(Pageable pageable);
}

File: 251003/Pankratiev/rest/publisher/src/main/kotlin/bsuir/dc/publisher/repository/IssueRepository.kt
package bsuir.dc.publisher.repository

import bsuir.dc.publisher.entity.Issue
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.JpaSpecificationExecutor
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param
import org.springframework.stereotype.Repository

@Repository
interface IssueRepository : JpaRepository<Issue, Long>, JpaSpecificationExecutor<Issue> {
    @Query(
        """
        SELECT DISTINCT i 
        FROM Issue i 
        JOIN i.labels l 
        WHERE l.id IN :labelIds 
        GROUP BY i 
        HAVING COUNT(DISTINCT l.id) = :labelCount
        """
    )
    fun findByAllLabelIds(@Param("labelIds") labelIds: List<Long>, @Param("labelCount") labelCount: Long): List<Issue>

    @Query(
        """
        SELECT DISTINCT i 
        FROM Issue i 
        JOIN i.labels l 
        WHERE l.name IN :labelNames 
        GROUP BY i 
        HAVING COUNT(DISTINCT l.name) = :labelCount
        """
    )
    fun findByAllLabelNames(@Param("labelNames") labelNames: List<String>, @Param("labelCount") labelCount: Long): List<Issue>
}
